drivers/sps30/sps30.c

```diff
@@ -0,0 +1,285 @@
+ *
+ * @return         SPS30_OK           if everything went fine
+ * @return         -SPS30_CRC_ERROR   if the CRC check failed
+ * @return         -SPS30_I2C_ERROR   if the I2C communication failed
+ */
+static int _rx_tx_data(const sps30_t *dev, uint16_t ptr_addr,
+                       uint8_t *data, size_t len, bool read)
+{
+    int res = 0;
+    unsigned retr = SPS30_ERROR_RETRY;
+
+    if (i2c_acquire(dev->p.i2c_dev) != 0) {
+        LOG_ERROR("could not acquire I2C bus %d\n", dev->p.i2c_dev);
+        return -SPS30_I2C_ERROR;
+    }
+
+    do {
+        size_t addr_data_crc_len = SPS30_PTR_LEN + len + len / 2;
+        uint8_t frame_data[addr_data_crc_len];
+        frame_data[0] = ptr_addr >> 8;
+        frame_data[1] = ptr_addr & 0xFF;
+
+        /* Both transfer types, `Set Pointer` and `Set Pointer & Read Data`
+           require writing a pointer address to the device in a separate write
+           transaction */
+        if (len == 0 || read) {
+            res = i2c_write_bytes(dev->p.i2c_dev, SPS30_I2C_ADDR,
+                                  &frame_data[0], SPS30_PTR_LEN, 0);
+        }
+
+        if (res == 0 && read) {
+            /* The `Set Pointer & Read Data` transfer type requires a separate
+               read transaction to actually read the data */
+            res = i2c_read_bytes(dev->p.i2c_dev, SPS30_I2C_ADDR,
+                                 &frame_data[SPS30_PTR_LEN],
+                                 addr_data_crc_len - SPS30_PTR_LEN, 0);
+
+            if (!_cpy_check_crc(data, len, &frame_data[SPS30_PTR_LEN])) {
+                res = -SPS30_CRC_ERROR;
+            }
+        } else {
+            /* For the `Set Pointer & Write Data` transfer type the full frame
+               is transmitted as one single chunk */
+            _cpy_add_crc(data, len, &frame_data[SPS30_PTR_LEN]);
+            res = i2c_write_bytes(dev->p.i2c_dev, SPS30_I2C_ADDR,
+                                  &frame_data[0], addr_data_crc_len, 0);
+        }
+    } while (res != 0 && retr--);
+
+    i2c_release(dev->p.i2c_dev);
+
+    return res == 0 ? SPS30_OK : -SPS30_I2C_ERROR;
+}
+
+int sps30_init(sps30_t *dev, const sps30_params_t *params)
+{
+    assert(dev && params);
+    dev->p = *params;
+    return sps30_start_measurement(dev);
+}
+
+int sps30_start_measurement(const sps30_t *dev)
+{
+    assert(dev);
+    uint8_t data[] = {SPS30_MESAURE_MODE, SPS30_DUMMY_BYTE};
+    return _rx_tx_data(dev, SPS30_CMD_START_MEASURE, (uint8_t*)data,
+                       sizeof(data), false);
+}
+
+int sps30_stop_measurement(const sps30_t *dev)
+{
+    assert(dev);
+    return _rx_tx_data(dev, SPS30_CMD_STOP_MEASURE, NULL, 0, false);
+}
+
+bool sps30_data_ready(const sps30_t *dev, int *error)
+{
+    assert(dev);
+    uint8_t data[2];
+    int res = _rx_tx_data(dev, SPS30_CMD_RD_DATA_READY, data, sizeof(data),
+                          true);
+    if (*error) {
+        *error = res;
+    }
+    return res == SPS30_OK ? data[1] : false;
```

Comment:
maybe `return (res == SPS30_OK) && data[1]` is even simpler here?

Reply:
I guess the compiler would have seen that too but I changed it anyway :)

Reply:
yep they are some smart these days